components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/logout/idpinit/processor/FederatedIdpInitLogoutProcessor.java

```diff
@@ -0,0 +1,632 @@
+            throw handleLogoutClientException(ErrorMessages.LOGOUT_TOKEN_PARSING_FAILURE, e);
+        } catch (JOSEException | IdentityOAuth2Exception e) {
+            throw handleLogoutServerException(ErrorMessages.LOGOUT_TOKEN_SIGNATURE_VALIDATION_FAILED, e);
+        }
+    }
+
+    /**
+     * Retrieve userId of the federated user.
+     *
+     * @param tenantDomain     - tenant domain of the logout request.
+     * @param sub              - sub claim in the logout token.
+     * @param identityProvider - identity provider.
+     * @return
+     * @throws LogoutServerException
+     */
+    private String getUserId(String tenantDomain, String sub, IdentityProvider identityProvider)
+            throws LogoutServerException {
+
+        try {
+            int tenantId = tenantDomain == null ? -1 : IdentityTenantUtil.getTenantId(tenantDomain);
+            String userId = UserSessionStore.getInstance()
+                    .getFederatedUserId(sub, tenantId, Integer.parseInt(identityProvider.getId()));
+            if (StringUtils.isBlank(userId)) {
+                throw handleLogoutServerException(ErrorMessages.RETRIEVING_USER_ID_FAILED, sub);
+            }
+            return userId;
+        } catch (UserSessionException e) {
+            throw handleLogoutServerException(ErrorMessages.RETRIEVING_USER_ID_FAILED, e, sub);
+        }
+    }
+
+    /**
+     * Check for iss claim in the logout token claims.
+     *
+     * @param claimsSet - claim set in the logout token.
+     * @return boolean.
+     */
+    private void isIssuerClaimExists(JWTClaimsSet claimsSet) throws LogoutClientException {
+
+        if (StringUtils.isBlank(claimsSet.getIssuer())) {
+            throw new LogoutClientException(ErrorMessages.LOGOUT_TOKEN_ISS_CLAIM_VALIDATION_FAILED.getCode(),
+                    ErrorMessages.LOGOUT_TOKEN_ISS_CLAIM_VALIDATION_FAILED.getMessage());
+        }
+    }
+
+    /**
+     * Do the aud claim validation according to OIDC back-channel logout specification.
+     *
+     * @param aud - list containing audience values.
+     * @param idp - identity provider.
+     * @return boolean if validation is successful.
+     */
+    private void validateAudience(List<String> aud, IdentityProvider idp) throws LogoutClientException {
+
+        String clientId = null;
+        // Get the client id from the authenticator config.
+        for (Property property : idp.getDefaultAuthenticatorConfig().getProperties()) {
+            String propertyName = property.getName();
+            if (propertyName.equals(OIDCAuthenticatorConstants.IdPConfParams.CLIENT_ID)) {
+                clientId = property.getValue();
+                break;
+            }
+        }
+        // Check whether the client id exist in the aud claim.
+        if (StringUtils.isNotBlank(clientId)) {
+            if (!aud.contains(clientId)) {
+                throw new LogoutClientException(ErrorMessages.LOGOUT_TOKEN_AUD_CLAIM_VALIDATION_FAILED.getCode(),
```

Comment:
If possible can we look to give a detailed message like

Audience validation failed. ClientID: X identifying the issuer: Y was not present as an audience value in the logout token.

or something along the lines